File: examples/provider_app/ios_virtual_wallet/VirtualWalletProvider/WalletPackage.swift
import AgentProvider
import Foundation

enum WalletPackage {
    static let providerId = "demo.virtual_wallet_provider"
    static let agentId = "demo.virtual_wallet.agent"

    static let actionPay = "wallet.payment.pay"
    static let actionListRecords = "wallet.records.list"
    static let actionConfigureQuietPay = "wallet.quiet_pay.configure"

    static let installURL = "wallet-provider://agent/install"
    static let actionURL = "wallet-provider://agent/action"
    static let hostInstallURL = "agent-host://agent-provider/add"
    static let hostTriggerURL = "agent-host://agent-provider/trigger"

    static let packageDef = AgentPackage(
        providerId: providerId,
        agentId: agentId,
        displayName: "Virtual Wallet Agent",
        description: "A local demo wallet for provider-confirmed and quiet small payments.",
        systemPrompt: """
        You help the user operate a virtual wallet through provider-owned actions.
        Use app_action_wallet_payment_pay when the user asks to pay a merchant.
        Use app_action_wallet_quiet_pay_configure when the user asks to enable, disable, or change small no-interruption payments.
        Use app_action_wallet_records_list when the user asks about recent spending.
        Payment is virtual demo data only, but still route payment proposals through the provider app.
        """,
        actions: [
            AgentAction(
                actionId: actionPay,
                toolName: "app_action_wallet_payment_pay",
                description: "Create a virtual payment record after provider policy and confirmation.",
                parameters: [
                    "type": .string("object"),
                    "properties": .object([
                        "merchant": .object(["type": .string("string")]),
                        "amount": .object([
                            "type": .string("number"),
                            "exclusiveMinimum": .number(0),
                        ]),
                        "currency": .object([
                            "type": .string("string"),
                            "default": .string("CNY"),
                        ]),
                        "note": .object(["type": .string("string")]),
                    ]),
                    "required": .array([.string("merchant"), .string("amount")]),
                ],
                resultSchema: resultSchema,
                risk: "high",
                confirmationPolicy: "provider_required",
                executionModes: ["app_handoff"],
                timeoutSeconds: 300
            ),
            AgentAction(
                actionId: actionListRecords,
                toolName: "app_action_wallet_records_list",
                description: "List recent virtual wallet payment records.",
                parameters: [
                    "type": .string("object"),
                    "properties": .object([
                        "limit": .object([
                            "type": .string("integer"),
                            "minimum": .number(1),
                            "maximum": .number(20),
                        ]),
                    ]),
                ],
                resultSchema: resultSchema,
                risk: "low",
                confirmationPolicy: "none",
                executionModes: ["app_handoff"],
                timeoutSeconds: 120
            ),
            AgentAction(
                actionId: actionConfigureQuietPay,
                toolName: "app_action_wallet_quiet_pay_configure",
                description: "Configure small no-interruption virtual payments.",
                parameters: [
                    "type": .string("object"),
                    "properties": .object([
                        "enabled": .object(["type": .string("boolean")]),
                        "limit": .object([
                            "type": .string("number"),
                            "minimum": .number(0),
                        ]),
                    ]),
                ],
                resultSchema: resultSchema,
                risk: "high",
                confirmationPolicy: "provider_required",
                executionModes: ["app_handoff"],
                timeoutSeconds: 300
            ),
        ],
        handoff: ["mode": .string("ios_url_handoff"), "display": .string("wallet_confirmation")],
        result: ["mode": .string("callback_url"), "schema": .string("wallet_result")]
    )

    private static let resultSchema: [String: JSONValue] = [
        "type": .string("object"),
        "properties": .object([
            "status": .object(["type": .string("string")]),
            "record": .object(["type": .string("object")]),
            "records": .object(["type": .string("array")]),
            "balance": .object(["type": .string("number")]),
            "balance_display": .object(["type": .string("string")]),
            "remaining_balance_text": .object(["type": .string("string")]),
            "quiet_pay_applied": .object(["type": .string("boolean")]),
            "message": .object(["type": .string("string")]),
        ]),
    ]
}

File: examples/provider_app/android_virtual_wallet/app/src/main/kotlin/demo/wallet/provider/WalletPackage.kt
package demo.wallet.provider

import agent.provider.sdk.AgentAction
import agent.provider.sdk.AgentPackage
import agent.provider.sdk.ConfirmationPolicy
import agent.provider.sdk.ExecutionMode

object WalletPackage {
    const val PROVIDER_ID = "demo.virtual_wallet_provider"
    const val AGENT_ID = "demo.virtual_wallet.agent"

    const val ACTION_PAY = "wallet.payment.pay"
    const val ACTION_LIST_RECORDS = "wallet.records.list"
    const val ACTION_CONFIGURE_QUIET_PAY = "wallet.quiet_pay.configure"

    val packageDef: AgentPackage
        get() = AgentPackage(
            providerId = PROVIDER_ID,
            agentId = AGENT_ID,
            displayName = "Virtual Wallet Agent",
            description = "A local demo wallet for provider-confirmed and quiet small payments.",
            systemPrompt = """
                You help the user operate a virtual wallet through provider-owned actions.
                Use app_action_wallet_payment_pay when the user asks to pay a merchant.
                Use app_action_wallet_quiet_pay_configure when the user asks to enable, disable, or change small no-interruption payments.
                Use app_action_wallet_records_list when the user asks about recent spending.
                Payment is virtual demo data only, but still route payment proposals through the provider app.
            """.trimIndent(),
            actions = listOf(
                AgentAction(
                    actionId = ACTION_PAY,
                    toolName = "app_action_wallet_payment_pay",
                    description = "Create a virtual payment record after provider policy and confirmation.",
                    parametersJson = paymentParameters,
                    resultSchemaJson = resultSchema,
                    risk = "high",
                    confirmationPolicy = ConfirmationPolicy.PROVIDER_REQUIRED,
                    executionModes = executionModes,
                    timeoutSeconds = 300,
                ),
                AgentAction(
                    actionId = ACTION_LIST_RECORDS,
                    toolName = "app_action_wallet_records_list",
                    description = "List recent virtual wallet payment records.",
                    parametersJson = """{"type":"object","properties":{"limit":{"type":"integer","minimum":1,"maximum":20}}}""",
                    resultSchemaJson = resultSchema,
                    risk = "low",
                    confirmationPolicy = ConfirmationPolicy.NONE,
                    executionModes = executionModes,
                    timeoutSeconds = 120,
                ),
                AgentAction(
                    actionId = ACTION_CONFIGURE_QUIET_PAY,
                    toolName = "app_action_wallet_quiet_pay_configure",
                    description = "Configure small no-interruption virtual payments.",
                    parametersJson = quietPayParameters,
                    resultSchemaJson = resultSchema,
                    risk = "high",
                    confirmationPolicy = ConfirmationPolicy.PROVIDER_REQUIRED,
                    executionModes = executionModes,
                    timeoutSeconds = 300,
                ),
            ),
            handoffJson = """{"mode":"android_activity_result","display":"wallet_confirmation"}""",
            resultJson = """{"mode":"activity_result","schema":"wallet_result"}""",
        )

    private val executionModes = listOf(
        ExecutionMode.APP_HANDOFF,
        ExecutionMode.ANDROID_ACTIVITY_RESULT,
    )

    private const val paymentParameters =
        """{"type":"object","properties":{"merchant":{"type":"string"},"amount":{"type":"number","exclusiveMinimum":0},"currency":{"type":"string","default":"CNY"},"note":{"type":"string"}},"required":["merchant","amount"]}"""

    private const val quietPayParameters =
        """{"type":"object","properties":{"enabled":{"type":"boolean"},"limit":{"type":"number","minimum":0}}}"""

    private const val resultSchema =
        """{"type":"object","properties":{"status":{"type":"string"},"record":{"type":"object"},"records":{"type":"array"},"balance":{"type":"number"},"balance_display":{"type":"string"},"remaining_balance_text":{"type":"string"},"quiet_pay_applied":{"type":"boolean"},"message":{"type":"string"}}}"""
}
